### ISSUE TITLE
Keydb Segfault

### ISSUE THREAD

**Issue by j4y4r** [JohnSully/KeyDB#109]
Hello,

we use keydb as cache for magento.
The key in question does not exist on the server,  the server only crashes when inserted to DB 0 this is where magento is writing to, this is not an issue on an empty database.

I could reproduce the issue when using the command below, the server crashes with a segfault.

``` keydb-cli -a "password" -h IP  HMSET "zc:k:792_BLOCK_FB44B6323D7A30F86CC92542AD343289F4E004FE" d "" t "792_STORE,792_CMS_B,792_BLOCK_HTML,792_MAGE" m "1574168516" i "0" ```

When i change the key to something else, like ``` zc:k:792_BLOCK_FB44B6323D7A30F86CC92542AD343289F4E004F1``` the server does not crash.

I checked the rdb, no errors where reported
output of  
keydb-check-rdb /srv/keydb/dump.rdb

[keydb_check_rdb_output.txt](https://github.com/JohnSully/KeyDB/files/3864956/keydb_check_rdb_output.txt)

full log
[keydb_log_output.txt](https://github.com/JohnSully/KeyDB/files/3864955/keydb_log_output.txt)

I tired with various Versions of Keydb, 
* latest build with  ``` docker run -it --rm -v /path-to-dump-binaries:/keydb_bin eqalpha/keydb-build-bin ```
* Redis server v=5.1.1 sha=a91f58b8:0 malloc=jemalloc-5.1.0 bits=64 build=535b18a2ba484f50, build from source
* Redis server v=5.1.1 sha=00000000:0 malloc=jemalloc-5.1.0 bits=64 build=685190c4b7a80c02, installed via the PPA
they all produce the same error with this database.

Also diffenent OS versions:

OS: Ubuntu 18.04 LTS
Kernel: Linux 4.15.0-20-generic

OS: Ubuntu 18.04.3 LTS
Kernel: Linux 4.15.0-70-generic

Keydb config:
/etc/keydb/keydb.conf

port 6379
requirepass pass
masterauth pass
active-replica yes
replicaof ip 6379
save 900 1
save 300 10k
save 60 10000
dir /srv/keydb/
dbfilename dump.rdb
syslog-enabled yes
syslog-ident keydb-server
pidfile /var/run/redis.pid

It makes no difference if the replication is running or not.

If you need any further imformation please let me know.

All The Best
Jeremias

**Comment by JohnSully** [comment]
It looks like this crash is dependent on the value currently stored under that key.  Can you provide a DUMP of the key "zc:k:792_BLOCK_FB44B6323D7A30F86CC92542AD343289F4E004FE"?

The assert is caused by us trying to add a new key that already exists, but I'm not yet sure how we came to make that decision.

**Comment by JohnSully** [comment]
Are you setting expires on the set?  Any subkey expires?

This case is very odd.  The high level key lookup is saying the key does not exist, but the low level hash table logic says it does.

**Comment by JohnSully** [comment]
I can successfully get into the state where this crashes.  It's caused by a key being expired but not actually flushed from the dictionary.  There are a few edge cases with expires and active replication that can get us here.  I should have a fix soon.

If you do not use expires let me know, as the bug I've found is predicated on this.

**Comment by JohnSully** [comment]
This is fixed in change: 8e4f323

Docker unstable is rebuilt nightly, the new build will automatically resolve the corruption in your database. This change will be in the next point release.

Note: There is now an incompatibility between old version active replicas and new ones. It is permissible to connect them for the purposes of an active upgrade but two versions should not be run simultaneously for any other purpose.

Active replicas are now responsible for performing their own expirations - this means an older server connected to a newer one will have stale keys kept around forever taking up memory (it will not show them to any clients). The server will warn about this condition, and in future versions servers without active expiration will not be allowed to connect to other active replicas.

**Comment by j4y4r** [comment]
Hey,

Just for completness:

asking the server for 
```keys "zc:k:792_BLOCK_FB44B6323D7A30F86CC92542AD343289F4E004F*"```
returns 
``` 
1) "zc:k:792_BLOCK_FB44B6323D7A30F86CC92542AD343289F4E004F2"
2) "zc:k:792_BLOCK_FB44B6323D7A30F86CC92542AD343289F4E004F1"
```

looking into the dump with 
```rdb --command json --key "zc:k:792_BLOCK_FB44B6323D7A30F86CC92542AD343289F4E004FE" dump.rdb```
returns
```
[{
"zc:k:792_BLOCK_FB44B6323D7A30F86CC92542AD343289F4E004FE":{"d":"","t":"792_STORE,792_CMS_B,792_BLOCK_HTML,792_MAGE","m":"1573699720","i":"0"}},{
"zc:k:792_BLOCK_FB44B6323D7A30F86CC92542AD343289F4E004FE":{"d":"","t":"792_STORE,792_CMS_B,792_BLOCK_HTML,792_MAGE","m":"1574168516","i":"0"}},{},{
"zc:k:792_BLOCK_FB44B6323D7A30F86CC92542AD343289F4E004FE":{"d":"","t":"792_STORE,792_CMS_B,792_BLOCK_HTML,792_MAGE","m":"1574168516","i":"0"}}]
```

Expires are set via the Application, but i do not know any specifics.

I build the unstable branch and it is in testing for approximatly 5 hours now, no issues so far :)
Thanks for the quick fix, Awesome !
We want to use Keydb in a Productive environment, can i do anything to help with the next point release?
Are there any info on when the next point release is scheduled?

Grettings
Jeremias

**Comment by JohnSully** [comment]
Hi @j4y4r 

There are a few more bugs tagged v5.2 I would like to get in the next release. Right now the unstable branch contains only bugfixes so it should be at release quality. 

If you are familiar with C++ I’m always looking for help with the core database. Otherwise your experiences with the current fix would be helpful.  Changing the Active Replication protocol is always challenging.

**Comment by JohnSully** [comment]
Specifically the blocker is #83 because that will also potentially require changes to Active Replication and I'd like to combine them in one release.

**Comment by JohnSully** [comment]
#83 was a bit easier than thought.  RELEASE_5 is staged while I do a bit more testing.  ETA is monday for the official release.

**Comment by j4y4r** [comment]
Hey,

Thanks for the information about the status of the unstable branch.

My C++ knowledge would be of no help i guess.

The Current buxfix, Redis server v=0.0.0 sha=39c6fbdd:0 malloc=jemalloc-5.1.0 bits=64 build=1e52cd3e0da53d9d, commit 39c6fbddb1dc6738c3ea7cbc2bd5d5bc7fb46ccf is running without issues since thursday.
Using Active Replication with 2 Servers with Ubuntu 18.04.3 LTS, Kernelversion 4.15.0-70-generic and the following configuration:

```
protected-mode yes
port 6379
tcp-backlog 511
timeout 0
tcp-keepalive 300
daemonize yes
supervised no
pidfile /var/run/keydb/keydb-server.pid
loglevel notice                        
logfile /var/log/keydb/keydb-server.log
syslog-enabled yes
syslog-ident keydb-server
databases 16
always-show-logo yes
save 900 1
save 300 10
save 60 10000
stop-writes-on-bgsave-error no
rdbcompression yes
rdbchecksum yes
dbfilename dump.rdb
dir /var/lib/keydb
replicaof ip 6379
masterauth password        
replica-serve-stale-data yes
repl-diskless-sync no
repl-diskless-sync-delay 5
repl-disable-tcp-nodelay no
replica-priority 100
requirepass password
lazyfree-lazy-eviction no
lazyfree-lazy-expire no
lazyfree-lazy-server del no
replica-lazy-flush no
appendonly no
appendfilename "appendonly.aof"
appendfsync everysec
no-appendfsync-on-rewrite no
auto-aof-rewrite-percentage 100
auto-aof-rewrite-min-size 64mb
aof-load-truncated yes
aof-use-rdb-preamble yes
lua-time-limit 5000
slowlog-log-slower-than 10000
slowlog-max-len 128
latency-monitor-threshold 0
notify-keyspace-events ""
hash-max-ziplist-entries 512
hash-max-ziplist-value 64
list-max-ziplist-size -2
list-compress-depth 0
set-max-intset-entries 512
zset-max-ziplist-entries 128
zset-max-ziplist-value 64
hll-sparse-max-bytes 3000
stream-node-max-bytes 4096
stream-node-max-entries 100
activerehashing yes
client-output-buffer-limit normal 0 0 0
client-output-buffer-limit replica 256mb 64mb 60
client-output-buffer-limit pubsub 32mb 8mb 60
hz 10
dynamic-hz yes
aof-rewrite-incremental-fsync yes
rdb-save-incremental-fsync yes
server-threads 2                                                             
active-replica yes
```

Looking forward to check out the next release.
If there is anything else i can to do to help, let me know.

Thanks for you support!

Kind Regards
Jeremias

**Comment by JohnSully** [comment]
Sorry for the long wait to publish 5.2, this month has been quite hectic.  However this is not officially released.

**Comment by j4y4r** [comment]
@JohnSully 
No worries ;)

In the release notes to 5.2 you state
"Fixed Issue #107 where expires are not always replicated with Active Replication"
the commit message refers to the wrong issue and so does the release note.

Thanks again for you quick fix and amazing support.

Fyi: still running the unstable branch without any issues :)

Greetings
Jeremias
